client/network/src/service.rs

```diff
@@ -844,6 +840,34 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 		}
 	}
 
+	/// Variation of `request` which can connect to peers when necessary.
+	///
+	/// In some cases we do request data, without prior notification. For those cases this function
+	/// exists. It has no guarantees on connection staying alive, so you might still want to pass
+	/// false to the `connect` parameter and ensure liveness of connections via peer sets.
+	/// Connecting to a peer can take up to several seconds, if a peer exposes multiple
+	/// addresses, but is only reachable by some of them, for example.
+	///
+	/// Apart from exposing whether or not we want to connect, in case of a disconnected peer, this
+	/// function also takes a sender for sending the response back, instead of returning the
+	/// response directly.
+	pub fn send_request(
+		&self,
+		target: PeerId,
+		protocol: impl Into<Cow<'static, str>>,
+		request: Vec<u8>,
+		tx: oneshot::Sender<Result<Vec<u8>, RequestFailure>>,
+		connect: bool,
```

Comment:
Instead of a `bool`, I'd suggest a new `IfDisconnectedBehaviour` enum, to make it clear what this does.
```suggestion
		connect: IfDisconnectedBehaviour,
```

```rust
pub enum IfDisconnectedBehaviour {
    TryConnect,
    ImmediateError,
}
```

You don't need to pass it around, it's just to make the public API clear. Feel free to pass a `bool` to the internal methods.

Reply:
I like this, was already tempted to do something like that. I would not mind changing the internal functions as well, I think this would improve readability over all.

Reply:
Ok, this would get noisy, due to the long name :-P Will probably stick with the public API for now.